File: src/main/java/com/itszuvalex/femtocraft/research/tiles/TileEntityResearchConsole.java
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */

package com.itszuvalex.femtocraft.research.tiles;

import com.itszuvalex.femtocraft.Femtocraft;
import com.itszuvalex.femtocraft.FemtocraftGuiHandler;
import com.itszuvalex.femtocraft.api.research.ITechnology;
import com.itszuvalex.femtocraft.api.research.ITechnologyCarrier;
import com.itszuvalex.femtocraft.core.tiles.TileEntityBase;
import com.itszuvalex.femtocraft.utils.FemtocraftDataUtils;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;

public class TileEntityResearchConsole extends TileEntityBase implements
        IInventory {
    private static final int ticksToResearch = 400;
    public
    @FemtocraftDataUtils.Saveable(desc = true)
    String displayTech = null;
    private
    @FemtocraftDataUtils.Saveable(desc = true)
    String researchingTech = null;
    private
    @FemtocraftDataUtils.Saveable
    int progress = 0;
    private
    @FemtocraftDataUtils.Saveable
    int progressMax = 0;
    private
    @FemtocraftDataUtils.Saveable
    ItemStack[] inventory = new ItemStack[10];

    public TileEntityResearchConsole() {
        super();
    }

    public int getResearchProgress() {
        return progress;
    }

    public void setResearchProgress(int progress) {
        this.progress = progress;
    }

    public int getResearchProgressScaled(int scale) {
        if (progressMax == 0) {
            return 0;
        }
        return (progress * scale) / progressMax;
    }

    public boolean isResearching() {
        return researchingTech != null;
    }

    public int getResearchMax() {
        return progressMax;
    }

    public void setResearchMax(int progressMax) {
        this.progressMax = progressMax;
    }

    public String getResearchingName() {
        return researchingTech;
    }

    @Override
    public void readFromNBT(NBTTagCompound par1nbtTagCompound) {
        super.readFromNBT(par1nbtTagCompound);
        checkForTechnology();
    }

    @Override
    public void writeToNBT(NBTTagCompound par1nbtTagCompound) {
        super.writeToNBT(par1nbtTagCompound);
    }

    @Override
    public void femtocraftServerUpdate() {
        super.femtocraftServerUpdate();
        if (researchingTech != null) {
            if (progress++ >= progressMax) {
                finishWork();
            }
        }
    }

    private void finishWork() {
        progress = 0;
        progressMax = 0;

        ITechnology tech = Femtocraft.researchManager()
                .getTechnology(researchingTech);
        if (tech == null) {
            researchingTech = null;
            return;
        }

        Item techItem = null;
        switch (tech.getLevel()) {
            case MACRO:
            case MICRO:
                techItem = Femtocraft.itemMicroTechnology();
                break;
            case NANO:
                techItem = Femtocraft.itemNanoTechnology();
                break;
            case DIMENSIONAL:
            case TEMPORAL:
            case FEMTO:
                techItem = Femtocraft.itemFemtoTechnology();
                break;
        }
        ItemStack techstack = new ItemStack(techItem, 1);
        ((ITechnologyCarrier) techItem).setTechnology(techstack,
                researchingTech);
        researchingTech = null;
        inventory[inventory.length - 1] = techstack;
        onInventoryChanged();
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
    }

    @Override
    public void markDirty() {
        super.markDirty();
        checkForTechnology();
    }

    private void checkForTechnology() {
        if (worldObj == null) {
            return;
        }
        if (worldObj.isRemote) {
            return;
        }

        boolean hadTech = displayTech != null;
        displayTech = null;

        for (ITechnology tech : Femtocraft.researchManager()
                .getTechnologies()) {
            if (Femtocraft.researchManager().hasPlayerDiscoveredTechnology(
                    getOwner(), tech) &&
                !Femtocraft.researchManager().hasPlayerResearchedTechnology(getOwner(), tech)) {
                if (matchesTechnology(tech)) {
                    displayTech = tech.getName();
                    if (worldObj != null) {
                        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
                    }
                    return;
                }
            }
        }

        if (hadTech && displayTech == null) {
            if (worldObj != null) {
                worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
            }
        }
    }

    private boolean matchesTechnology(ITechnology tech) {
        if (tech == null) {
            return false;
        }
        if (tech.getResearchMaterials() == null || tech.getResearchMaterials().length == 0) {
            return false;
        }
        for (int i = 0; i < 9 && i < tech.getResearchMaterials().length; ++i) {
            if (!compareItemStack(tech.getResearchMaterials()[i], inventory[i])) {
                return false;
            }
        }

        return true;
    }

    private boolean compareItemStack(ItemStack tech, ItemStack inv) {
        return tech == null && inv == null || tech != null && inv != null && tech.stackSize <= inv.stackSize &&
                                              ItemStack.areItemStacksEqual(tech, inv);

    }


    /*
     * (non-Javadoc)
     *
     * @see TileEntityBase#hasGUI()
     */
    @Override
    public boolean hasGUI() {
        return true;
    }

    @Override
    public int getGuiID() {
        return FemtocraftGuiHandler.ResearchConsoleGuiID();
    }

    public void startWork() {
        if (!canWork()) {
            return;
        }

        researchingTech = displayTech;
        progressMax = ticksToResearch;
        progress = 0;

        ITechnology tech = Femtocraft.researchManager()
                .getTechnology(displayTech);

        for (int i = 0; i < 9 && i < tech.getResearchMaterials().length; ++i) {
            if (tech.getResearchMaterials()[i] == null) {
                continue;
            }
            decrStackSize(i, tech.getResearchMaterials()[i].stackSize);
        }

        this.onInventoryChanged();
    }

    private boolean canWork() {
        if (displayTech == null || displayTech.isEmpty()) {
            return false;
        }
        if (researchingTech != null && !researchingTech.isEmpty()) {
            return false;
        }
        ITechnology tech = Femtocraft.researchManager()
                .getTechnology(displayTech);

        return matchesTechnology(tech);
    }

    @Override
    public int getSizeInventory() {
        return inventory.length;
    }

    @Override
    public ItemStack getStackInSlot(int i) {
        return inventory[i];
    }

    @Override
    public ItemStack decrStackSize(int i, int j) {
        if (this.inventory[i] != null) {
            ItemStack itemstack;

            if (this.inventory[i].stackSize <= j) {
                itemstack = this.inventory[i];
                this.inventory[i] = null;
                return itemstack;
            } else {
                itemstack = this.inventory[i].splitStack(j);

                if (this.inventory[i].stackSize == 0) {
                    this.inventory[i] = null;
                }

                return itemstack;
            }
        } else {
            return null;
        }
    }

    @Override
    public ItemStack getStackInSlotOnClosing(int i) {
        return inventory[i];
    }

    @Override
    public void setInventorySlotContents(int i, ItemStack itemstack) {
        inventory[i] = itemstack;

    }

    @Override
    public String getInventoryName() {
        return "Research Console";
    }

    @Override
    public boolean hasCustomInventoryName() {
        return false;
    }

    @Override
    public int getInventoryStackLimit() {
        return 64;
    }

    @Override
    public boolean isUseableByPlayer(EntityPlayer entityplayer) {
        return canPlayerUse(entityplayer);
    }

    @Override
    public void openInventory() {
    }

    @Override
    public void closeInventory() {
    }

    @Override
    public boolean isItemValidForSlot(int i, ItemStack itemstack) {
        switch (i) {
            case 9:
                return false;
            default:
                return true;
        }
    }

}

File: src/main/scala/com/itszuvalex/femtocraft/research/tiles/TileEntityResearchConsole.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.research.tiles

import com.itszuvalex.femtocraft.api.EnumTechLevel._
import com.itszuvalex.femtocraft.api.core.Saveable
import com.itszuvalex.femtocraft.api.items.ITechnologyCarrier
import com.itszuvalex.femtocraft.api.research.ITechnology
import com.itszuvalex.femtocraft.core.tiles.TileEntityBase
import com.itszuvalex.femtocraft.core.traits.tile.Inventory
import com.itszuvalex.femtocraft.utils.BaseInventory
import com.itszuvalex.femtocraft.{Femtocraft, FemtocraftGuiConstants}
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.oredict.OreDictionary

import scala.collection.JavaConversions._

object TileEntityResearchConsole {
  private val ticksToResearch: Int = 400
}

class TileEntityResearchConsole extends TileEntityBase with Inventory {
  @Saveable(desc = true) var displayTech: String = null
  @Saveable(desc = true) private var researchingTech: String = null
  @Saveable private              var progress                = 0
  @Saveable private              var progressMax             = 0


  override def defaultInventory = new BaseInventory(10)

  def getResearchProgress = progress

  def setResearchProgress(progress: Int) {
    this.progress = progress
  }

  def getResearchProgressScaled(scale: Int): Int = {
    if (progressMax == 0) {
      return 0
    }
    (progress * scale) / progressMax
  }

  def isResearching = researchingTech != null

  def getResearchMax = progressMax

  def setResearchMax(progressMax: Int) {
    this.progressMax = progressMax
  }

  def getResearchingName = researchingTech

  override def readFromNBT(par1nbtTagCompound: NBTTagCompound) {
    super.readFromNBT(par1nbtTagCompound)
    checkForTechnology()
  }

  private def checkForTechnology() {
    if (worldObj == null) {
      return
    }
    if (worldObj.isRemote) {
      return
    }
    val hadTech = displayTech != null
    displayTech = null
    Femtocraft.researchManager.getTechnologies.
    filter(Femtocraft.researchManager.hasPlayerDiscoveredTechnology(getOwner, _)).
    filterNot(Femtocraft.researchManager.hasPlayerResearchedTechnology(getOwner, _)).
    filter(matchesTechnology).
    foreach { tech =>
      displayTech = tech.getName
      if (worldObj != null) worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
      return
            }
    if (hadTech && displayTech == null) {
      if (worldObj != null) {
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
      }
    }
  }

  private def matchesTechnology(tech: ITechnology): Boolean = {
    if (tech == null) {
      return false
    }
    if (tech.getResearchMaterials == null || tech.getResearchMaterials.length == 0) {
      return false
    }
    for (i <- 0 until Math.min(9, tech.getResearchMaterials.length)) {
      if (!compareItemStack(tech.getResearchMaterials(i), getStackInSlot(i))) {
        return false
      }
    }
    true
  }

  private def compareItemStack(tech: ItemStack, inv: ItemStack): Boolean = {
    tech == null && inv == null || tech != null && inv != null && tech.stackSize <= inv.stackSize && tech.getItem == inv.getItem && (tech.getItemDamage == inv.getItemDamage || tech.getItemDamage == OreDictionary.WILDCARD_VALUE || inv.getItemDamage == OreDictionary.WILDCARD_VALUE)
  }

  override def writeToNBT(par1nbtTagCompound: NBTTagCompound) {
    super.writeToNBT(par1nbtTagCompound)
  }

  override def femtocraftServerUpdate() {
    super.femtocraftServerUpdate()
    if (researchingTech != null) {
      if ( {progress += 1; progress - 1} >= progressMax) {
        finishWork()
      }
    }
  }

  private def finishWork() {
    progress = 0
    progressMax = 0
    val tech = Femtocraft.researchManager.getTechnology(researchingTech)
    if (tech == null) {
      researchingTech = null
      return
    }
    var techItem: Item = null
    tech.getLevel match {
      case MACRO | MICRO                  =>
        techItem = Femtocraft.itemMicroTechnology
      case NANO                           =>
        techItem = Femtocraft.itemNanoTechnology
      case FEMTO | DIMENSIONAL | TEMPORAL =>
        techItem = Femtocraft.itemFemtoTechnology
    }
    val techstack = new ItemStack(techItem, 1)
    techItem.asInstanceOf[ITechnologyCarrier].setTechnology(techstack, researchingTech)
    researchingTech = null
    setInventorySlotContents(getSizeInventory - 1, techstack)
    markDirty()
    worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
  }

  override def markDirty() {
    super.markDirty()
    checkForTechnology()
  }

  override def hasGUI = true

  override def getGuiID = FemtocraftGuiConstants.ResearchConsoleGuiID

  def startWork() {
    if (!canWork) {
      return
    }
    researchingTech = displayTech
    progressMax = TileEntityResearchConsole.ticksToResearch
    progress = 0
    val tech = Femtocraft.researchManager.getTechnology(displayTech)
    for (i <- 0 until Math.min(9, tech.getResearchMaterials.length)) {
      if (tech.getResearchMaterials(i) != null) {
        decrStackSize(i, tech.getResearchMaterials(i).stackSize)
      }
    }
    markDirty()
  }

  override def getInventoryName = "Research Console"

  override def hasCustomInventoryName = false

  override def isItemValidForSlot(i: Int, itemstack: ItemStack) = i match {
    case 9 => false
    case _ => true
  }

  private def canWork: Boolean = {
    if (displayTech == null || displayTech.isEmpty) {
      return false
    }
    if (researchingTech != null && !researchingTech.isEmpty) {
      return false
    }
    val tech = Femtocraft.researchManager.getTechnology(displayTech)
    matchesTechnology(tech)
  }
}
